src/handlers/web3.ts

```diff
@@ -0,0 +1,679 @@
+ * @return The corresponding `TransactionResponse`, or `null` if one could not
+ * be found.
+ */
+export const getTransaction = async (
+  hash: string
+): Promise<TransactionResponse | null> =>
+  web3Provider ? web3Provider.getTransaction(hash) : null;
+
+/**
+ * @desc get address transaction count
+ * @param address The address to check.
+ * @return The transaction count, or `null` if it could not be found.
+ */
+export const getTransactionCount = async (
+  address: string
+): Promise<number | null> =>
+  web3Provider ? web3Provider.getTransactionCount(address, 'pending') : null;
+
+/**
+ * @desc Gets transaction details for a new transaction.
+ * @param transaction The new transaction. In particular, the `from`, `to`,
+ * `gasPrice`, `gasLimit`, `amount` fields from a `NewTransaction` are required,
+ * as well as an optional `data` field similar to a `TransactionRequest`.
+ * @return The transaction details.
+ */
+export const getTxDetails = async (
+  transaction: Partial<
+    Pick<
+      NewTransactionNonNullable,
+      'from' | 'to' | 'gasPrice' | 'gasLimit' | 'amount'
+    >
+  > &
+    Pick<TransactionRequest, 'data'>
+): Promise<TransactionDetailsReturned> => {
+  const { to } = transaction;
+  const data = transaction.data ? transaction.data : '0x';
+  const value = transaction.amount ? toHex(toWei(transaction.amount)) : '0x00';
+  const gasLimit = transaction.gasLimit
+    ? toHex(transaction.gasLimit)
+    : undefined;
+  const gasPrice = transaction.gasPrice
+    ? toHex(transaction.gasPrice)
+    : undefined;
+  const tx = {
+    data,
+    gasLimit,
+    gasPrice,
+    to,
+    value,
+  };
+  return tx;
+};
+
+/**
+ * @desc Resolves an Unstoppable domain string.
+ * @param domain The domain as a string.
+ * @returns The resolved address, or undefined if none could be found.
+ */
+export const resolveUnstoppableDomain = async (
+  domain: string
+): Promise<string | void> => {
+  const resolution = new UnstoppableResolution({
+    blockchain: {
+      cns: {
+        network: 'mainnet',
+        url: rpcEndpoints[Network.mainnet],
+      },
+    },
+  } as any);
+
+  const res = resolution
+    .addr(domain, 'ETH')
+    .then(address => {
+      return address;
+    })
+    .catch(logger.error);
+  return res;
+};
+
+/**
+ * @desc Resolves a name or address to an Ethereum hex-formatted address.
+ * @param nameOrAddress The name or address to resolve.
+ * @param provider If provided, a provider to use instead of the cached
+ * `web3Provider`.
+ * @returns The address, or undefined if one could not be resolved.
+ */
+export const resolveNameOrAddress = async (
+  nameOrAddress: string,
+  provider: JsonRpcProvider | null = null
+): Promise<string | void> => {
+  if (!isHexString(nameOrAddress)) {
+    if (/^([\w-]+\.)+(crypto)$/.test(nameOrAddress)) {
+      return resolveUnstoppableDomain(nameOrAddress);
+    }
+    const p = provider || web3Provider;
+    return p?.resolveName(nameOrAddress);
```

Comment:
This shouldn't need the optional chaining anymore — guessing this is based on an earlier type for `web3Provider`?

Reply:
I think it might make sense to keep this because `web3Provider` is defined as a `JsonRpcProvider`, but actually could still be `null`: https://github.com/rainbow-me/rainbow/pull/2511#discussion_r741602383. In order to keep compatibility with other modules, I think it makes sense to keep that non-null definition, but it also means that it's safer to do optional checks where possible. Let me know what you think, thanks!

Reply:
I raised this because the original code is `return p.resolveName(nameOrAddress)` which would throw an error if it was ever `null`, but it certainly doesn't hurt to guard here either, so I'm happy either way!